### DISCUSSION TITLE
IdP selection based on login email domain

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#8839] [Ideas]
We have a request for a feature which selects the target Identity provider based on the login email address of the user. Instead of a list of login buttons, the user will get just an email input field and a text like "Please provide your email, we then open the appropriate login page for you". The feature needs a mapping configuration with entries which map an email domain names to target IdP. It also needs to deliver a new login screen where just the email of the user is requested.

The feature is relevant in scenarios where you have multiple IdPs configured at one realm where users of different companies login with their business email as username. You don't want to end up with users selecting the wrong IdP and you also don't want to present the complete list of configured IdPs at the login screen. Several cloud IdPs apply this pattern, like e.g. MS Azure at https://portal.azure.com/.

The feature may be implemented with a custom Authenticator. When having it open source it could be included in the default Authentication flow like the default "Identity Provider Redirector" (class `IdentityProviderAuthenticator`). Just deactivated, without any mappings configured. A soon as you configure one or more domain to IdP mapping, the screen which requests the user's email would be displayed. Such a new `ResolveIdPAuthenticator` could be included right after the existing "Identity Provider Redirector".

Is there any interest in such an Authenticator developed and contributed to the Keycloak Open Source project?

**Comment by anshumanchatterji** [comment]
Within a month of this FR raised, I need this feature as well. When you have a subdomain for every customer, you can’t create those many clients and configure different IDPs for them.

**Comment by stianst** [comment]
Red Hat also has this need. Right now it's being handled in a bunch of extensions and custom flows, but we'd love to get this into Keycloak.

To extend on this a bit though we have the need to define a group of users associated with a customer, that is linked to a particular IdP. This will also force this group of users to login with the IdP. We also want to link this up with LDAP and/or SCIM to support syncing of users. Which further starts going down to route of a sub-realm within a realm to have a context for a specific customer, where other things can live as well, like groups, roles, service accounts, etc.

@mposolda

**Reply by stianst** [reply]
@alechenninger FIY

**Reply by alechenninger** [reply]
Yeah the more I think about it the more it seems like a first-class "tenant" within a realm is going to simplify a lot. And then of course, given a "realm" already models a tenant, reusing a realm as in "subrealms" seems natural. Attempts to get around using a subrealm seem to be awkward and error prone to me. Should we take an honest look at a couple of options of what it would take to get a subrealm concept off the ground in the most iterative way possible?

**Reply by jbman** [reply]
I think sub-realm level tenants would be another great feature but not a precondition for this idea.
The idea as described currently and as implemented by @sventorben at https://github.com/sventorben/keycloak-home-idp-discovery selects the IdP based on the email domain. Of cause later the Authenticator could be extended to map based on some tenant the user belongs to, but for now email domain would be sufficient.

**Reply by alechenninger** [reply]
I agree subrealms are not a precondition to purely IdP selection based on login email domain by itself. I have a [very similar but hacky proof of concept that does the domain-based redirect as well](https://github.com/alechenninger/keycloak/tree/federated-idp-by-domain), and we do some related things on redhat.com.

At the end of the day, when you have this scenario of different subsets of users belonging to different customers (in your particular scenario this customer also owns a particular domain), I think customers want:

* A tenant which they can self-service manage the things they care about
* ...but reusing existing clients which are already integrated with the Keycloak instance (so this is not a totally separate realm)
* Ability to prove ownership of a domain and associate it with that tenant (so the customer's new users are automatically registered within their tenant if they have that domain for their email). Note this is useful even without also federating IdPs.
* Ability to link the lifecycle of their users, their groups, and their roles with their own, already well-managed identity system by way of IdP federation and/or data syncing/data federation using SCIM or LDAP. In other words, they don't want to have to onboard or disable users when they join their company or get added to some group in another IdP when they already built tools to manage that in their organization's IdP.
* Ability to define some policy about their tenant like password complexity, password rotation, or additional authentication factors, etc. based on their needs (e.g. in order to comply with regulations that apply to them)

You can tackle just one feature in isolation, but how do the rest of the features come together? Do they end up with a sensible solution to the customer and Keycloak administrator–one easy to manage abstraction? Or are they disjointed, requiring additional extensions or heavy refactoring to pull them together? Honest questions, but I have my doubts.

Even just trying to get domain-based IdP right can get tricky in isolation, for example:

* If a customer already has a set of existing users (say, under a certain group or identified with a certain attribute), you'd probably have to have something else which ensured these users all used the same email domain in order to be redirected to the IdP for that customer. Or if you can't do something to lock down the email domain you'd need a second check on top of the email domain, like a check based on a group or attribute.
* Users can also freely remove an IdP link by default, which is another reason you may need a secondary check or a way to prevent changing their email to an email with a different domain, or a policy about the users which prevents them from removing that IdP link.
* You have to make sure the identity provider redirector is not further up in the authenticator flow, because if it's an alternative, users could use that to pick an alternative IdP and still login (like a social provider), so you also need a way to prevent that. This also assumes other IdPs are not just available on the login page to use in the first place.
* You have to also be sure that no clients are allowed to use resource owner password credential grants for these users or prevent users from configuring passwords. Otherwise, users can still configure passwords and get around the IdP requirement via the ROPC grant.
* Since a password reset flow logs in a user at the end normally, you have to also make sure users with such an IdP link can't get around it by doing a password reset

This is all to say, an authenticator alone isn't a complete solution. Several other pieces have to come together. And it still doesn't yet provide the other features the customer is looking for. First class federation at a "tenant" level (e.g. subrealm, but perhaps a group could work with some caveats, open to ideas there of course), would be able to address the whole solution with a few basic config options that encapsulate all of that complexity rather than requiring many different configurations to come together in exactly the right way to provide the solution.

There is a related proposal we're working on a Red Hat that may be a smaller stepping stone, I'll link to that here when it is shared. I'm just wondering if we should maybe start pursuing subrealm or something like it sooner than later.

**Comment by sventorben** [comment]
@jbman I think I have what you are looking for right [here](https://github.com/sventorben/keycloak-home-idp-discovery) as an extension.
I haven't invested much time into it, yet. So, it works but is not complete. It's missing some admin UI.

I could contribute this to the Keycloak project if desired and if it does not need too much work for additional features.

**Reply by jbman** [reply]
@sventorben that looks very promising and very much like suggested in this idea. I also like the convenient configuration directly at the identity provider. However is requires this extra management endpoint for the Identity providers (HomeIdpDiscoveryRealmResource.java) With mapping configuration at the Authenticator it would be a more cohesive implementation but a less convenient to manage.

Would appreciate if it can go into Keylcloak. Maybe maintainers could also give a feedback?

**Reply by sventorben** [reply]
@jbman It does not necessarily need that `HomeIdpDiscoveryRealmResource` endpoint. When storing the config at the identity provider, there is already a default endpoint for that within Keycloak (see the [README](https://github.com/sventorben/keycloak-home-idp-discovery/#configure-email-domains))
Initially I wanted to store the configuration somewhere else. So, you may think of that endpoint as a kind of "legacy".

I would suggest to add a list of domains to the [`IdentityProviderModel`](https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/models/IdentityProviderModel.java) instead. [This wrapper](https://github.com/sventorben/keycloak-home-idp-discovery/blob/main/src/main/java/de/sventorben/keycloak/authentication/hidpd/IdentityProviderModelConfig.java) would become obsolete in that case. In addition, we could have a list of domains in the indentity provider UI tab similiar to valid redirect URIs for clients.

**Reply by mnaser** [reply]
We're using this extension which has solved us with that problem, but now even if we have a "Hide on Login Page" option selected, unfortunately it still shows up in the account console "Linked Accounts", so I think we'd need a "Hide on Account Console" option in there too.

**Reply by mnaser** [reply]
I just created #13206 to track this :)

**Reply by sventorben** [reply]
Do you use the new account console already? With the new console, it should be easy to remove the "Linked IdP section" from the `content.json` file. However, this will hide all Identity Providers.

**Reply by mnaser** [reply]
@sventorben we do use it, but then that will indeed kill off the ability to see current links with IdP's such as Google/GitHub/etc.

**Reply by nitdgpvijay** [reply]
@jbman  I deployed your extension on kyecloak 19.0.0, but dont see "Home Idp discovery" tab on "identity providers -> provider details" page.
Although routing works based on email domain, I can update domain via rest api.
Also I would need to hide list of IDP on login page.
Could you please help here.
![image](https://user-images.githubusercontent.com/25844895/182808015-13f12b66-d096-4c39-9e56-606c82c9d41b.png)

**Reply by sventorben** [reply]
@nitdgpvijay The new admin console is not supported yet.
https://github.com/sventorben/keycloak-home-idp-discovery/issues/86
